logging fix and the long-awaited setting log-level by processor

tomeichlersmith opened the pull request:
These developments do the following
- interpret the message's logging level into a human readable string for easier parsing
- add event number to logging messages via a custom Formatter which resolves #1301 
- allow logging levels to differ between different channels by implementing a custom Filter which resolves #922 
- update the configuration so that the logging has its own configuration class, keeping its parameters organized and separate from the Process (the current syntax is supported by overriding `__getattr__` so this doesn't require folks to update their configs).
- only printout an event status timestamp on the first try of every event which resolves #1455  

to lower the level for everyone
```python
p.logger.termLevel = 0
# OR, with legacy syntax
p.termLogLevel = 0
```
to debug a specific processor
```python
p.logger.debug(my_processor)
# OR
p.logger.debug("MyProcessorName")
```
to silence a specific processor
```python
p.logger.silence(my_processor)
# OR
p.logger.silence("MyProcessorName")
```
or something in between
```python
p.logger.custom(my_processor, level = 1)
```

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.

These developments were tested with a simple standalong processor and config that I've copied here for completeness.

```python
from LDMX.Framework import ldmxcfg
p = ldmxcfg.Process('log')
p.sequence = [ldmxcfg.Producer.from_file('EgLog.cxx')]
p.outputFiles = [ '/dev/null' ]
p.maxEvents = 5
p.logFrequency = 1
p.termLogLevel = 0
p.logger.debug(p.sequence[0])
```

```cpp
#include "Framework/EventProcessor.h"

class EgLog : public framework::Producer {
 public:
  EgLog(const std::string& name, framework::Process& p)
    : framework::Producer(name, p) {}
  virtual ~EgLog() = default;
  void produce(framework::Event& event) override {
    ldmx_log(info) << "Info message";
    ldmx_log(error) << "Error message";
  }
};

DECLARE_PRODUCER(EgLog);
```

tomeichlersmith committed:
put event number into logging messages

also convert the logging levels to human readable strings while we are working on a formatter anyways

tomeichlersmith committed:
promote logging into its own Python configuration class

this will be more helpful as I attempt to add the additional features of
per-processor log filtering

tomeichlersmith committed:
add custom logging rules allowing specific channels to be raised or
lowered

tomeichlersmith committed:
only print event message on first tries

tomeichlersmith committed:
just format-cpp

tomeichlersmith commented:
first round of PR Validations: https://github.com/LDMX-Software/ldmx-sw/actions/runs/10890162256/job/30218290716

tvami commented on Framework/src/Framework/Logger.cxx:
```diff
@@ -116,6 +142,47 @@ void close() {
+      os << "debug";
+      break;
+    case level::info:
+      os << "info ";
+      break;
+    case level::warn:
+      os << "warn ";
```
Extremely minor suggestion, but speaking of parsing, it could be important
```suggestion
      os << "info";
      break;
    case level::warn:
      os << "warn";
```

tomeichlersmith replied:
So I left the spaces in on purpose so that the message label is constant width. Do we want to keep that constant width? (And maybe apply a width to the event number and processor names) or not worry about constant width

tvami replied:
I see, well with the `debug` the printout is `debug:` while with the info that becomes `info :` which looked a bit weird. And indeed with the processor name being not the same length I didnt even think of the reason for having a space is the same length. I dont think it's realistic to have a fixed length for the processors, sometimes we have very long names, so I'd just give up on this.

tvami replied:
I guess it's true that it does make it nicer inside a given processor
```
[ ElectronCounter ] 9838 debug: Found 1 electrons (tracks) using input collection TriggerPadTracksY_
[ trigger ] 9838 debug: Got trigger energy cut 1500 for 1 electrons counted in the event.
[ trigger ] 9838 debug: Got trigger energy sum 966.672; and decision is pass = true
[ ecalVeto ] 9839 debug: Straight tracks found (before merge): 0
[ ecalVeto ] 9839 debug: Beginning track merging using 0 tracks
[ ecalVeto ] 9839 debug: Straight tracks found (after merge): 0
[ ecalVeto ] 9839 debug: Finding linreg tracks from 4 hits
[ ecalVeto ] 9839 info :   MIP tracking completed; found 0 straight tracks and 0 lin-reg tracks
[ ecalVeto ] 9839 debug:   The pred > bdtCutVal = true
[ ElectronCounter ] 9839 debug: Found 1 electrons (tracks) using input collection TriggerPadTracksY_
```

tvami replied:
So what if we put the space in front, like ` info` and ` warn`? That will satisfy both the fixed length, and will not introduce a space between the text and the colon

bryngemark replied:
sounds like an easy way out

tvami reviewed (changes requested):
Awesome work resolving several issues! I have a small comment below, otherwise great!

bryngemark commented on Framework/src/Framework/Process.cxx:
```diff
@@ -484,8 +481,10 @@ void Process::newRun(ldmx::RunHeader &header) {
+bool Process::process(int n, int n_try, Event &event) const {
+  if ((logFrequency_ != -1) && ((n + 1) % logFrequency_ == 0) && (n_try < 2)) {
+    // only printout event counter if we've enabled log frequency, the event
+    // matches the frequency and we are on the first try
```
this change reminds me, i've grown used to having to set the termLogLevel to 1 or less to get event number printouts even with logFrequency set to something reasonable (like 10 printouts per job). does that change in the current implementation?

tomeichlersmith replied:
This does not; however, I could implement something to avoid that.

From the logging's perspective, both processors and the `Process` object (and anything that logs) are just "channels" that are sending logging messages. I could lower the Process channel's log level if a logFrequency_ > 0 is provided so that the Process's messages are displayed.

bryngemark reviewed (approved):
thanks for finding the time to do this, @tomeichlersmith. long-awaited indeed :) 

when this is in, am i right that we should go through all processors (maybe not that many! but most of mine) which have a separate verbosity functionality and move them to only using the logger's severity?

tomeichlersmith commented:
> when this is in, am i right that we should go through all processors (maybe not that many! but most of mine) which have a separate verbosity functionality and move them to only using the logger's severity?

That is correct @bryngemark ! Instead of `verbose` flags I would imagine that these messages would be put into `debug` and then people could opt-in to view the debug messages for certain processors with
```python
p.logger.debug(processor)
```

I'm open to hearing other ideas on the user interface for this. While I like having the Process and Logger configuration separate on the C++ side, the Python side is very flexible and (as can be seen with `__setattr__`), we can get pretty :magic_wand: magical :magic_wand: with it.

tomeichlersmith committed:
move space to before label so colon always follows label

also use a const reference instead of a copy just because we can lmao

tomeichlersmith committed:
lower the level of the Process channel to info if a logFrequency is set

this avoids tying together the event-status frequency from the overall
terminal log level

tvami reviewed (approved):
Perfect, thanks!

tomeichlersmith commented:
So the `ecal_pn` sample is failing to run

```
---- LDMXSW: Loading configuration --------
  Traceback (most recent call last):
    File "config.py", line 179, in <module>
  Setting detector for tracking to ldmx-det-v14-8gev-no-cals
      recoil_dqm.measurement_collection=digi_recoil.out_collection
  NameError: name 'digi_recoil' is not defined. Did you mean: 'digiRecoil'?
  Configuration Error [ConfigureError] : Problem running python script.
    at /home/runner/work/ldmx-sw/ldmx-sw/Framework/src/Framework/ConfigurePython.cxx:319 in ConfigurePython
```

which was introduced in #1454 

I'm guessing this is what is being referred to by
> we updated the tracking configs

in #1459

tvami commented:
> I'm guessing this is what is being referred to by

I confirm that, I already have a working version in the branch, but we should have https://github.com/LDMX-Software/ldmx-sw/pull/1460 merged and then I can do the other items in that issue and push it
